library(shiny)
library(ggvis)


options(shiny.table.class = "data table table-striped table-condensed ")

shinyUI(fluidPage(
  
  tags$head(
    includeCSS("web/styles1.css")
  ),
  
  tags$head(
    includeScript("web/ggcircos.js")
  ),
  
  # App title
  titlePanel(HTML("<h3>exomePeak2 - visualizer</h3>"),windowTitle = "exomePeak2 visualizer"),

  
  fluidRow(
  column(7,
         
         div(id = "app_info", class = "collapse out", 
             p("This visualisation is based on Circos, a means of visualising whole genomes. The goal here is to visualize the peak calling results on MeRIP-seq data generated by exomePeak2 - an R/Bioconductor package."),             
             tags$a(href = "https://bioconductor.org/packages/release/bioc/html/exomePeak2.html"),
             p("Although this plot cannot give much biological information, it's useful for brief view of analysis outputs. Please refer to UCSC Genome Browser or other visualization tools where you can obtain informative knowledge related to biological functions."),
             br(),
             p("This plot shows a number of layers. Firstly, the lines represent peaks along the whole genome in each chromosome. By clicking on one chromosome block, the peak regions are revealed. You can control how much the chromosome expands by using Resize Factor slider below the table. The bars upwards are peaks on the positive strands, and the bars downwards are peaks on the negative strands. Hover the bars for specific ranges of the peaks."),
             p("Moving inwards, adjusted p-values are shown as a scatter plot. Red points are adjusted p-values less than 0.05, which reveals that the differential expression of this region/genes is significant. Hover the points for more details. ")
             ),
         
         
         
         HTML("<button type='button' class='btn' data-toggle='collapse' style='float:left' data-target='#app_info'><span class='glyphicon glyphicon-collapse-down'></span> More Information</button>"),
         
         br(),  br(), 
   
         ggvisOutput("plot")

  ),
  
  column(5,
         br(),

         div(id = "genes_div"),
 
         div(id = "transcripts_div"),

         div(id = "clinvar_div"),
         
         h4("Summary of Peak Calling Results (Top10 significant):"),
         
         tableOutput("top_clinvar_genes"),
         
         br(),   
         
         
         
         sliderInput("integer", "Resize Factor:", min = 1, max = 15, value = 5)
         
  )),
  fluidRow(column(12,
                  tags$br(),
                  hr(),
                  p("App created by Di Zhen in Feb 2021", HTML("&bull;"), "Find the code on Github:", tags$a(href = "https://github.com/JoKerDii/m6a-seq-analysis-visualizer", "JoKerDii", tags$i(class = 'fa fa-github', style = 'color:#5000a5'), target = '_blank'), style = "font-size: 80%"),
                  p(tags$em("Last updated: Feb 2021"), style = 'font-size:75%')))
)
)